### PR TITLE
[FIX] base: attachment copy ACL

### PR DESCRIPTION
**Pull request by xmo-odoo**
Currently, `Attachment.copy` incurs an *explicit* check for write access on the underlying record.

This doesn't necessarily make sense e.g. a user copying an attachment from a template to an email may not have write access to the template, but that should not be an issue. And indeed if the copy is performed "by hand" (read then create) things work fine.

Since both `read` and `create` are checked, the explicit `copy` check doesn't seem necessary. Drop it, and try to add some tests around `copy`.

A secondary issue is that the `check` in `create` was performed in the context of the `self` being copied, leading to the same issue as `copy` being repeated (that is, `a.copy()` would call `a.create()` which would call `a.check('write')`, even though we're semantically creating a record from scratch). The behavior is really intended for `write` where we want to check if we have `write` access to both the "source" and the "destination" records.

Explicitly opt-out of having any source data in the `create` before performing the `check` calls, this way we correctly and only check for the validity of the destination.

issue 2746483

### CODE REVIEW

**Review by KangOl — APPROVED** [review]

File: odoo/addons/base/tests/test_ir_attachment.py [hunk 1]
```diff
@@ -75,53 +75,58 @@ def test_06_linked_record_permission(self):
...
+        a3 = a1.copy({'res_model': writable3._name, 'res_id': writable3.id})
+
         # can copy to self with write permission
-        a2_copy.copy()
-        # can't copy back to record without write permission
+        a3.copy()
+        # cannot copy back to record without write permission
         with self.assertRaises(AccessError):
-            a2_copy.copy({'res_id': vals['res_id']})
+            a3.copy({'res_model': record1._name, 'res_id': record1.id})
```
Comment (xmo-odoo) [inline odoo/addons/base/tests/test_ir_attachment.py:127]: This rather misses the point, which was copying cross records without even *touching* the model, to verify that that behaves correctly.

The behaviour when the model is explicit is already checked at lines 119 and 121.

**Review by rco-odoo — APPROVED** [review]
That's good to me, but the original test was unreadable, and Xavier's additions made it even more unreadable.  So I pushed an extra commit to make the code of the test more readable.  I leave it up to you to squash it to your commit, if you find it okay.

File: odoo/addons/base/tests/test_ir_attachment.py [hunk 2]
```diff
@@ -75,53 +75,58 @@ def test_06_linked_record_permission(self):
...
-        a3 = Attachment.create({'name': 'yes'})
-        a3.name = 'canwrite'
-        a2_copy = a2.copy({'res_id': a3.id})
+        a3 = a1.copy({'res_model': writable3._name, 'res_id': writable3.id})
+
         # can copy to self with write permission
-        a2_copy.copy()
-        # can't copy back to record without write permission
+        a3.copy()
+        # cannot copy back to record without write permission
```
Comment (xmo-odoo) [inline odoo/addons/base/tests/test_ir_attachment.py:125]: This replacement makes no sense. Both are contractions of "can not" and "cannot" is both longer and less popular than "can't".

File: odoo/addons/base/tests/test_ir_attachment.py [hunk 3]
```diff
@@ -75,53 +75,58 @@ def test_06_linked_record_permission(self):
...
+        vals = {'name': 'attach', 'res_id': record1.id, 'res_model': record1._name}
+        a1 = Attachment.create(vals)
 
-        # remove access to linked record a1
+        # remove all access rights to record1
         rule = self.env['ir.rule'].create({
-            'name': 'test_rule', 'domain_force': "[('id', '!=', %s)]" % a1.id,
-            'model_id': self.env.ref('base.model_ir_attachment').id,
+            'name': 'test_rule', 'domain_force': "[('id', '!=', %s)]" % record1.id,
+            'model_id': self.env['ir.model']._get_id(record1._name),
```
Comment (xmo-odoo) [inline odoo/addons/base/tests/test_ir_attachment.py:86]: The goal was not to rewrite the entire test in 13, it was to add checks for previously untested behaviours.

**Comment by xmo-odoo** [comment]
Applied suggestions.

Since we were largely rewriting the specific test for readability, and so much was being modified for the new properties being checked, I ended up splitting it in a different test case with clearer individual tests.

A possible improvement would be to make the "copy" test paraetrized / table-driven so it's a bit more formal and regular. But that seemed a bit much.

**Comment by rco-odoo** [comment]
I just pushed a mini cleanup (avoid the single-line `with` block).
The test is way much better this way :+1:

**Review by rco-odoo — APPROVED** [review]

**Comment by rco-odoo** [comment]
@robodoo r+

**Pull request closed without merging**